src/common/pico_sync/sem.c

```diff
@@ -52,6 +52,17 @@ bool __time_critical_func(sem_acquire_block_until)(semaphore_t *sem, absolute_ti
     } while (true);
 }
 
+bool __time_critical_func(sem_try_acquire)(semaphore_t *sem) {
+    uint32_t save = spin_lock_blocking(sem->core.spin_lock);
+    if (sem->permits == 0) {
+        spin_unlock(sem->core.spin_lock, save);
+        return false;
+    }
+    sem->permits--;
```

Comment:
Looking at the implementations of `sem_acquire_blocking` and `sem_acquire_block_until`, should this check that `sem->permits > 0` before decrementing it?

Reply:
Good point, I think it would be better to reverse the logic -- check for `permits > 0`, if so then acquire the permit and return true. Otherwise release the lock and return false.

Reply:
@lurch or @kilograham if you agree I'll make this change, say the word.

Reply:
sounds good